File: .claude/skills/performance-audit/test-fixtures/go-sample/inventory.go
package shop

import "fmt"

type Item struct {
	ID    string
	Name  string
	Price int
}

type Quote struct {
	Total int
}

type Totals struct {
	Revenue  int
	Tax      int
	Shipping int
}

// FindDuplicateSKUs returns SKUs that appear more than once.
//
// PLANTED #4 (algorithmic): membership test against a SLICE (`contains`) inside
// the loop is O(n) per check → O(n^2) overall. Use a map[string]struct{} set.
// Request-sized input on a hot path.
func FindDuplicateSKUs(skus []string) []string {
	var seen []string
	var dupes []string
	for _, sku := range skus {
		if contains(seen, sku) { // O(n) linear scan inside the loop
			dupes = append(dupes, sku)
		} else {
			seen = append(seen, sku)
		}
	}
	return dupes
}

func contains(xs []string, x string) bool {
	for _, v := range xs {
		if v == x {
			return true
		}
	}
	return false
}

// BuildLabels formats a label per item.
//
// PLANTED #5 (memory): `labels` grows by append from a nil slice with no
// preallocation — repeated doublings + copies. `make([]string, 0, len(items))`
// pre-sizes it.
//
// BEYOND-THE-PACK (floor-not-ceiling): `fmt.Sprintf("%d", it.Price)` to convert
// an int to a string on a hot path uses reflection and is ~an order of magnitude
// slower than `strconv.Itoa(it.Price)`. NO Go-pack bullet names fmt.Sprintf-for-
// int-conversion; the agent must know/reason that fmt is reflection-based here.
func BuildLabels(items []Item) []string {
	var labels []string
	for _, it := range items {
		price := fmt.Sprintf("%d", it.Price)
		labels = append(labels, it.Name+": "+price)
	}
	return labels
}

// defaultRegions is a fixed 3-element config read once at startup.
var defaultRegions = []string{"us", "eu", "apac"}

// IsSupportedRegion — DECOY: `contains` over a SLICE, which mirrors the O(n^2)
// pattern, BUT defaultRegions is a constant of 3 and this is a single membership
// test (not nested in a request loop). O(3) is not a finding; flagging "use a map"
// here is checklist-walking.
func IsSupportedRegion(region string) bool {
	return contains(defaultRegions, region)
}

func (s *Server) fetchRevenue(orderID string) (int, error)  { return 0, nil }
func (s *Server) fetchTax(orderID string) (int, error)      { return 0, nil }
func (s *Server) fetchShipping(orderID string) (int, error) { return 0, nil }

File: .claude/skills/performance-audit/test-fixtures/python-sample/inventory.py
"""Inventory operations. Called on the order-processing hot path.

Contains two planted performance issues (see test-fixtures/.../expected-findings.md).
"""

import repo


def find_duplicate_skus(skus):
    """Return SKUs that appear more than once.

    PLANTED ISSUE #1 (Lane 1 — algorithmic): membership test `in seen` against a
    LIST inside the loop is O(n) per check → O(n^2) overall. `seen` should be a set.
    Reached per request with request-sized `skus`.
    """
    seen = []
    dupes = []
    for sku in skus:
        if sku in seen:          # O(n) scan of a list, inside a loop
            dupes.append(sku)
        else:
            seen.append(sku)
    return dupes


def enrich_line_items(order_item_ids):
    """Attach catalog data to each line item in an order.

    PLANTED ISSUE #2 (Lane 3 — data access / N+1): one repo.get() call per item
    inside the loop. repo.get_many() can fetch the whole batch in a single
    round-trip. Reached per order on the checkout path.
    """
    enriched = []
    for item_id in order_item_ids:
        row = repo.get(item_id)   # N+1: one round-trip per item
        if row:
            enriched.append({"id": item_id, "name": row["name"], "price": row["price"]})
    return enriched
